[13.0][ADD] sale_order_qty_change_no_recompute: Prevent recompute if only quantity has changed in sale order line-

victoralmau opened the pull request:
Prevent recompute if only quantity has changed in sale order line.

Please @pedrobaeza and @carlosdauden can you review it?

@Tecnativa TT29404

victoralmau committed:
[ADD] sale_order_qty_change_no_recompute: Prevent recompute if only quantity has changed in sale order line

youring commented:
Functional tested not work on [runbot.](http://3480769-1573-60ceb9.runbot1.odoo-community.org/?)
Change the unit price, then change the quantity will still trigger recomputing.

pedrobaeza commented on sale_order_qty_change_no_recompute/models/sale_order.py:
```diff
@@ -0,0 +1,21 @@
+    )
+    def _onchange_discount(self):
+        if (
+            self.product_uom_qty == self._origin.product_uom_qty
```
This condition is not correct

victoralmau replied:
Condition are changed now.

pedrobaeza commented on sale_order_qty_change_no_recompute/models/sale_order.py:
```diff
@@ -0,0 +1,21 @@
+    @api.onchange(
+        "product_id", "price_unit", "product_uom", "product_uom_qty", "tax_id"
+    )
+    def _onchange_discount(self):
```
Are you sure this is the proper method to override?

victoralmau replied:
Now addon override only `product_uom_change` method.

pedrobaeza reviewed (changes requested):
Please generate README

HaraldPanten reviewed (changes requested):
I've been testing in runbot, but everytime the quantity is changed, prices are recomputed at sale order line level.

youring commented:
https://github.com/OCA/sale-workflow/commit/dcdef5c4616abd91270ef975cb3ca1b8361a5ad6 Works as expected now.

victoralmau commented:
> I've been testing in runbot, but everytime the quantity is changed, prices are recomputed at sale order line level.

Sorry, it's necessary save order before try to change quantity and prevent prices recomputation.

pedrobaeza committed:
[FIX] sale_order_qty_change_no_recompute: Proper way for removing onchange

The _origin trick is not valid, as it doesn't reflect the state before
the onchange, only the saved copy (if any).

pedrobaeza commented on sale_order_qty_change_no_recompute/tests/test_sale_order_qty_change.py:
```diff
@@ -0,0 +1,51 @@
+        cls.partner = cls.env.ref("base.res_partner_12")
+
+    def test_sale_line_without_change(self):
+        sale_form = Form(self.env["sale.order"])
```
Put this in `setUpClass`.

pedrobaeza commented on sale_order_qty_change_no_recompute/tests/test_sale_order_qty_change.py:
```diff
@@ -0,0 +1,46 @@
+            self.assertEqual(line_form.price_subtotal, 25)
+
+    def test_sale_line_change_price(self):
+        with self.sale_form.order_line.new() as line_form:
```
I think you can even put the line_form and the product set on the setUpClass

pedrobaeza commented on sale_order_qty_change_no_recompute/tests/test_sale_order_qty_change.py:
```diff
@@ -0,0 +1,47 @@
+
+    def test_sale_line_change_price(self):
+        self._set_init_line()
+        self.line_form.price_unit = 10
```
I don't get the sense of this test. Changing the price unit is not what this module prevents.

pedrobaeza commented on sale_order_qty_change_no_recompute/tests/test_sale_order_qty_change.py:
```diff
@@ -0,0 +1,47 @@
+        with sale_form.order_line.new() as cls.line_form:
+            cls._set_init_line(cls)
+
+    def _set_init_line(self):
```
Why doing this sub-method? In fact the only thing you should test is:

- Fix a price on the line
- Change product > check the price changes
- Change quantity > check that the price doesn't change.

pedrobaeza commented:
@HaraldPanten can you please update your review?

youring commented:
Functional tested on proper way. LG!
But I like the naming of this module by @StefanRijnhart , more accessible for functional user.
https://github.com/opener-am/various-odoo-addons/tree/8.0/sale_keep_unit_price_on_qty_change
Just FYI.

pedrobaeza commented:
That name is not accurate by several reasons:

- It doesn't include `order` for specifying in which part of `sale` this happens.
- The computation can affect also the discount, not only the price unit.

Anyway, there's no good name most of the times, but you have the README for explaining the exact goal ;) 

![imagen](https://user-images.githubusercontent.com/7165771/118237810-f47fa700-b497-11eb-8e4f-55c43361fa3f.png)

pedrobaeza commented:
Can you approve then the PR if you agree?

HaraldPanten reviewed (approved):
LGTM. Sorry @pedrobaeza , I didn't read you before.

pedrobaeza commented:
/ocabot merge nobump

incoluz commented:
Hello, this is not working for us on 14.0-20220606 core version. Only prevents when uom changes not on qty change.
